atomix/cluster/src/main/java/io/atomix/raft/storage/log/RaftLogReader.java

```diff
@@ -18,50 +18,122 @@
 
 import io.atomix.raft.storage.log.entry.RaftLogEntry;
 import io.atomix.storage.journal.Indexed;
-import io.atomix.storage.journal.JournalReader;
+import io.zeebe.journal.JournalReader;
+import io.zeebe.journal.JournalRecord;
+import java.nio.ByteBuffer;
+import java.util.NoSuchElementException;
+import org.agrona.DirectBuffer;
 
 /** Raft log reader. */
 public class RaftLogReader implements java.util.Iterator<Indexed<RaftLogEntry>>, AutoCloseable {
-  private final JournalReader<RaftLogEntry> delegate;
+  private final RaftLog log;
+  private final JournalReader journalReader;
+  private final RaftLogReader.Mode mode;
 
-  RaftLogReader(final JournalReader<RaftLogEntry> delegate) {
-    this.delegate = delegate;
-  }
-
-  public long getFirstIndex() {
-    return delegate.getFirstIndex();
-  }
+  // NOTE: nextIndex is only used if the reader is in commit mode, hence why it's not subject to
+  // inconsistencies when the log is truncated/compacted/etc.
+  private long nextIndex;
 
-  public long getLastIndex() {
-    return delegate.getLastIndex();
+  RaftLogReader(
+      final RaftLog log, final JournalReader journalReader, final RaftLogReader.Mode mode) {
+    this.log = log;
+    this.journalReader = journalReader;
+    this.mode = mode;
   }
 
   @Override
   public boolean hasNext() {
-    return delegate.hasNext();
+    return (mode == Mode.ALL || nextIndex <= log.getCommitIndex()) && journalReader.hasNext();
```

Comment:
`nextIndex` is not initialized. Could that be a problem? If I check the usage, immediately after creating the reader, it is reset to a specific index. So I guess it is ok.

Reply:
It will be 0 by default, which should be less than the commit index if we've committed anything ever, and equal to the commit index if we've never committed anything. So we should probably initialize it to the first index :+1:

Reply:
This is what is causing RandomizedTest to fail. Initially both nextIndex and commitIndex = 0.  So then CommittedReader still returns true for hasNext(), if there is atleast one uncommitted entry.